storage: return LeaseRejectedError when learner requests lease

tbg opened the pull request:
We would previously return an unstructured error which would leak to the
client.
The LeaseRejectedError gets turned into a NotLeaseholderError instead,
and DistSender will try another replica.

Fixes #40323.

Note that DistSender doesn't send to learners in the first place, so
it's unclear why we asked a learner to request a lease, but it's likely
that we upgraded the learner to a voter while it was down (the test
below restarts nodes) and it received a request while it was already
a voter but not aware of it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40329)
<!-- Reviewable:end -->

tbg committed:
storage: return LeaseRejectedError when learner requests lease

We would previously return an unstructured error which would leak to the
client.
The LeaseRejectedError gets turned into a NotLeaseholderError instead,
and DistSender will try another replica.

Fixes #40323.

Note that DistSender doesn't send to learners in the first place, so
it's unclear why we asked a learner to request a lease, but it's likely
that we upgraded the learner to a voter while it was down (the test
below restarts nodes) and it received a request while it was already
a voter but not aware of it.

Release note: None

tbg commented:
Fixed the test failure.

bors r=danhhz
